### PR TITLE
Add CAN rgb led

### PR DESCRIPTION
**Pull request by CUAVcaijie**
Add CAN rgb led

### CODE REVIEW

File: src/drivers/uavcan/sensors/rgb.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2014, 2015 PX4 Development Team. All rights reserved.
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.cpp:3]:
```suggestion
 *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```

File: src/drivers/uavcan/sensors/rgb.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	_timer(node)
+{
+
+}
+
+int
+UavcanUavcanRgb::init()
+{
+	/*
+	 * Setup timer and call back function for periodic updates
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.cpp:55]: Can you move the description to the header file? Thanks

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/****************************************************************************
+*
+*   Copyright (c) 2014, 2015 PX4 Development Team. All rights reserved.
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.hpp:3]:
```suggestion
*   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+****************************************************************************/
+
+/**
+ * @file rgb.hpp
+ *
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.hpp:36]: Can you add a description with `@brief` annotation? Thanks

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+#include <uavcan/uavcan.hpp>
+#include <uavcan/equipment/indication/LightsCommand.hpp>
+#include <perf/perf_counter.h>
+#include <lib/led/led.h>
+
+/**
+ * @brief The RGB class
+ */
+
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.hpp:50]:
```suggestion
```

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	/*
+	* setup periodic updater
+	*/
+	int init();
+
+private:
+	/*
+	 * Max update rate to avoid exessive bus traffic
+	 */
+	static constexpr unsigned			MAX_RATE_HZ = 100;
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.hpp:65]:
```suggestion
	static constexpr unsigned MAX_RATE_HZ = 100;
```

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+	void periodic_update(const uavcan::TimerEvent &);
+
+	typedef uavcan::MethodBinder<UavcanUavcanRgb *, void (UavcanUavcanRgb::*)(const uavcan::TimerEvent &)>
+	TimerCbBinder;
+	LedController		_led_controller;
+
+	/*
+	 * libuavcan related things
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.hpp:74]: Please replace with a description of each object.

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	/*
+	 * Max update rate to avoid exessive bus traffic
+	 */
+	static constexpr unsigned			MAX_RATE_HZ = 100;
+
+	void periodic_update(const uavcan::TimerEvent &);
+
+	typedef uavcan::MethodBinder<UavcanUavcanRgb *, void (UavcanUavcanRgb::*)(const uavcan::TimerEvent &)>
+	TimerCbBinder;
+	LedController		_led_controller;
```
Comment (TSC21) [inline src/drivers/uavcan/sensors/rgb.hpp:71]:
```suggestion
	LedController _led_controller;
```

File: src/drivers/uavcan/sensors/rgb.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+ * @file rgb.cpp
+ *
+ * @author CUAVcaijie <caijie@cuav.net>
+ */
+
+#include "rgb.hpp"
+#include <systemlib/err.h>
+
+
+UavcanUavcanRgb::UavcanUavcanRgb(uavcan::INode &node) :
```
Comment (JacobCrabill) [inline src/drivers/uavcan/sensors/rgb.cpp:44]: Can we name this something like `UavcanRgbLed` instead of having `UavcanUavcan` in the name?

Comment (CUAVcaijie) [reply src/drivers/uavcan/sensors/rgb.cpp:44]: Ok

File: src/drivers/uavcan/sensors/rgb.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+		case led_control_s::COLOR_GREEN:
+			_r = 0; _g = 255; _b = 0;
+			break;
+
+		case led_control_s::COLOR_BLUE:
+			_r = 0; _g = 0; _b = 255;
+			break;
+
+		case led_control_s::COLOR_AMBER: //make it the same as yellow
```
Comment (JacobCrabill) [inline src/drivers/uavcan/sensors/rgb.cpp:88]:
```suggestion
		case led_control_s::COLOR_AMBER: // make it the same as yellow
```

File: src/drivers/uavcan/sensors/rgb.hpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+	/*
+	 * libuavcan related things
+	 */
+	uavcan::Publisher<uavcan::equipment::indication::LightsCommand> _rgb_pub;
+	uavcan::TimerEventForwarder<TimerCbBinder>			_timer;
```
Comment (JacobCrabill) [inline src/drivers/uavcan/sensors/rgb.hpp:77]:
```suggestion
	uavcan::TimerEventForwarder<TimerCbBinder> _timer;
```

File: src/drivers/uavcan/sensors/rgb.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	if (!_timer.isRunning()) {
+		_timer.setCallback(TimerCbBinder(this, &UavcanUavcanRgb::periodic_update));
+		_timer.startPeriodic(uavcan::MonotonicDuration::fromMSec(1000 / MAX_RATE_HZ));
+	}
+
+	return 0;
+}
+
+void
+UavcanUavcanRgb::periodic_update(const uavcan::TimerEvent &)
```
Comment (JacobCrabill) [inline src/drivers/uavcan/sensors/rgb.cpp:66]:
```suggestion
void UavcanUavcanRgb::periodic_update(const uavcan::TimerEvent &)
```

Comment (JacobCrabill) [reply src/drivers/uavcan/sensors/rgb.cpp:66]: I think I've seen the general PX4 style moving away from the multi-line function header...

File: src/drivers/uavcan/sensors/rgb.hpp (hunk 6 above)
Comment (bkueng) [inline src/drivers/uavcan/sensors/rgb.hpp:65]: 100 Hz still seems excessive. What about 20 or 30?

Comment (CUAVcaijie) [reply src/drivers/uavcan/sensors/rgb.hpp:65]: 20Hz is also ok , the break light is not so good

Comment (CUAVcaijie) [reply src/drivers/uavcan/sensors/rgb.hpp:65]: 20 is also is ok , the effect of the breathing light will not be particularly smooth

**Comment by dagar** [comment]
> Maybe @dagar has an opinion on this as well?

Yes, but if we're going to do anything here I'd probably want to start by making it a bit more dynamic (at startup). We're paying for a lot of additional state for every possible use case.

**Comment by dagar** [comment]
There's a built failure that needs to be fixed, I'll take a look.

**Comment by CUAVcaijie** [comment]
Thank you

**Comment by AlexandreBorowczyk** [comment]
I've tried this PR my branch based on 1.10.1 running on the V5 Nano and I've notice that the board's led won't light up.
Stoping rgbled_pwm and starting it again will fix the issue but will to crash the UavcanNode.

Also should _led_ be added to the _DEPENDS_ in the CMakeLists?

**Pull request closed without merging**